src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc

```diff
@@ -833,4 +709,190 @@ BtreeLBAManager::_update_mapping(
     });
 }
 
+BtreeLBAManager::alloc_extents_ret
+BtreeLBAManager::_alloc_extents(
+    Transaction &t,
+    laddr_hint_t hint,
+    std::vector<alloc_mapping_info_t> &alloc_infos)
```

Comment:
IIUC, let's take the chance to simplify the pre-conditions and add explicit comments, for example:

```
Allocate a list of pins.
* If hint isn't null, allocate a consecutive pins of the total length based on the hint (alloc_mapping_info_t::key must be null).
* If hint is null, allocate a list of pins based on the fixed non-null alloc_mapping_info_t::key.
```

IIUC there are only 2 use cases above, please correct me if wrong.

Reply:
After re-examining the code, I think it's better to keep the hint always non-null, as it will be passed to `search_insert_pos`. Otherwise, we should construct the hint.addr from the first entry of `alloc_infos`.

Having the caller construct the hint seems more reasonable.

---

Maybe add an assertion `laddr.hint == alloc_infos.fonrt().key` when the key is not null?

Reply:
I might be missing some contexts above, after another look, does this pre-condition look accurate now?

```
Allocate a list of pins:
1. If alloc_mapping_info_t::key are all null, allocate a list of consecutive pins of their total length based on the hint.
2. If alloc_mapping_info_t::key are all not-null, allocate a list of adjacent pins from these fixed keys, and the hint must be fixed (never conflict) covering the correct range.
3. alloc_mapping_info_t::key must be all null or not-null.
```

Reply:
Also need to add an assertion `hint.policy == laddr_conflict_policy_t::never` for case 2.